### DISCUSSION TITLE
Vehicle availability time constraint (shifts) prevent seeing pickup times

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#2837] [Routing (and legacy CP) questions] [answered]
Hi,

I am using or-tools successfully for a VRPTW problem with pickup and deliveries. On top of classic time constraints at each pickup and delivery locations, I have a constraint that some vehicle different start and end times availability. I have solved it by imposing constraint:
```cs
solver.Add(solver.MakeGreaterOrEqual(timeDimension.CumulVar(this.RoutingModel.Start(nVehicle)), vehicle_available_startTime - 15));
solver.Add(solver.MakeLessOrEqual(timeDimension.CumulVar(this.RoutingModel.End(nVehicle)), vehicle_available_endTime + 15));
```
The idea to set a constraint that vehicle nVehicle cannot start pickups before vehicle_available_startTime  (e.g. 10am) and 
cannot make deliveries after vehicle_available_endTime (e.g. 4pm) with a slack of 15 minutes before or after.
**BTW: is this the right way of doing this?** This seems to work fine, vehicles indeed start pickups and make last deliveries as defined by the constraint aobve.

PROBLEM: when I go to print the solution, the value of min and max for the solution of each vehicle at each node is returned as the vehicles start time! Though the solution seems correct! What I mean is:
```cs
for (int i = 0; i < this.DriverVehicles.Count; ++i)
{
   var routeNodeIndex = RoutingModel.Start(i);
   var timeVar = timeDimension.CumulVar(routeNodeIndex);
   Console.Write("{0} Time({1},{2}) -> ", manager.IndexToNode(index), solution.Min(timeVar), solution.Max(timeVar));
}
```
code above prints 
Vehicle 0:
0--> 1 Time (140, 140)  -> 2 Time (140, 140) -> 3 Time (140, 140) -> 0 Time (140,140)
(140 is start time of vehicle 0)

Vehicle 1:
0--> 4 Time (160, 160)  -> 5 Time (160, 160) -> 6 Time (160, 160) -> 0 Time (160,160)
(160 is start time of vehicle 1)

where the solution is correct, that's the proper way of visiting nodes, but 140 is the availability start time of vehicle 0 and 160 is availability start time of vehicle 1

Any idea why is this happening? **I need to see the actual windows when each nodes is visited.** 

Thanks!

**Comment by Mizux** [comment, marked as answer]
May I have the full print output function, you should have two loops:
```
for each vehicle v:
  index = RoutingModel.Start(v)
  for index up to index ==  RoutingModel.End(v)
     ....
```

also in your snippet you are using `manager.IndexToNode(index)` but we can't see any `index` declared...

**Reply by odashevsky** [reply]
Hi,

Sorry, I skipped a in showing my code. I do indeed have inner loop as you mentioned. But I found the silly bug: I did not update "timeVar" inside the inner loop, doh!

Thanks for your help.

**Reply by akilash93** [reply]
Can you please share the complete working solution for this? @odashevsky Trying to work on a similar problem and would need help on the implementation.
